### ISSUE TITLE
Angular Sample - XML supplied to Blockly.inject does not appear in browser

### ISSUE THREAD

**Issue by christopherallanperry** [RaspberryPiFoundation/blockly-samples#50]
### TLDR
`Blockly.inject()` doesn't provide XML to the browser in either the Angular sample, or my own implementation based on it.

### Concept Summary
I've built out a version of the Blockly Sample for Angular with a view to getting the [Code Editor](https://blockly-demo.appspot.com/static/demos/code/index.html) working inside it. In theory, this shouldn't require a lot more than injecting the appropriate XML, adding the `code.js`, and including some additional structure to the component HTML.

### Angular Implementation
To implement this, I've created an Angular Service, calling a separate XML file, to supply the Blockly XML into the component - `blockly.component.ts` looks like this.

![blockly_component_ts](https://user-images.githubusercontent.com/10407622/72983813-20c27080-3dda-11ea-9a11-89bbd4e6b417.PNG)

### Chrome DevTools Error
What I'm seeing though is an error in the console, where line 405 of `code.js` says it can't find `id="toolbox"`. 

![code_js_405_error](https://user-images.githubusercontent.com/10407622/72982920-14d5af00-3dd8-11ea-8e0e-c7dc106ad220.PNG)

Looking at the HTML in the browser, it appears that the injected XML is missing from `blocklyDiv`, and all I'm seeing is a structure based on SVG elements.

![no_xml_blockly](https://user-images.githubusercontent.com/10407622/72983007-46e71100-3dd8-11ea-9e12-ed4c04b69e0f.PNG)
### Rendered App
In the browser window, this is what I'm seeing rendered. As you can see, the  category names are not populating, due to the console error.

![blockly_code_app_screen](https://user-images.githubusercontent.com/10407622/72983137-91688d80-3dd8-11ea-8282-8797bfa677a0.PNG)
### Current Workaround
The only way I've found to address the problem, is to render out the XML directly using...
`
const blocklyDiv = document.getElementById("blocklyDiv");
blocklyDiv.innerHTML = xml;
`
...where `xml` is a reference to the XML string itself, and cutting out the `Blockly.inject()` method altogether, which feels wrong and I think likely to catch me out later when I come to the point where I need to add custom blocks into the mix.

### Google Blockly Sample - Angular
I've gone back to a clean copy of the Angular sample to understand what the behavior is there. Again, there is no sign of the XML structure, just the SVG one - and no `id="toolbox"` in the HTML.

![clean_blockly_sample_angular](https://user-images.githubusercontent.com/10407622/72984654-27ea7e00-3ddc-11ea-8ac2-adc152e98f56.PNG)

At this point, I'm not sure whether I'm implementing things incorrectly, or whether there is a bug associated with `Blockly.inject()` when used in Angular.

Any suggestions?

**Comment by christopherallanperry** [comment]
### Additional Observations
Whilst working on this a bit more, I've noticed what I assume are side effects of the `Blockly.inject()` method not being in play, the following don't work on my instance

1. The scroll bars are inactive
2. I'm unable to drag blocks around the workspace
3. The wheel doesn't zoom in/out of the workspace.

**Comment by samelhusseini** [comment]
If you pass the toolbox in ``Blockly.inject`` as a string, then Blockly would parse that xml string. If you pass the toolbox as an xml string, then it's not injected anywhere in the browser.
Think of it as Blockly having two ways of reading toolbox xml:
- the xml is already injected (by you) in the browser, and you pass us the XML Element (Node)
- the xml you're passing is an xml string, and we can parse that string ourselves.
See https://github.com/google/blockly/blob/master/core/options.js#L304 for details.

The Blockly Angular sample takes the second approach.

Hard to say what's going on with your code without seeing the xml service, but if it returns a string then it needs to be a parsable xml string by ``Blockly.xml.textToDom``, otherwise it must return an xml dom element.

Here's a working example of using a dom element for the toolbox instead of a string: 
https://github.com/google/blockly-samples/compare/angular_toolbox_dom

**Comment by christopherallanperry** [comment]
Thanks for all that Sam,

It does rather sound like I'm passing it in the wrong format. I'll have a look a bit later and see if that does the trick.

If you're interested, my repo can be found here - https://github.com/christopherallanperry/ng_blockly

And thanks again.

**Comment by christopherallanperry** [comment]
I've just pushed an updated version to my repo based on turning the output of the Blockly service into an XML DOM structure, then injected that into `Blockly.inject()`. 

Again, without using `blocklyDiv.innerHTML = xml` to set up the DOM, I'm seeing the same errors. Note: the version I've just pushed has all that commented out for the moment, so you should see what I do if you spin it up.

Grateful for any additional suggestions at this point.

**Comment by samelhusseini** [comment]
Took a quick look at your repo, and it looks like there's a few things interfering with each other. 
The angular sample injects a blockly workspace into the #blocklyDiv div, but then you have the code demo in there which injects a workspace in the #content_blocks div.

The toolbox XML specifies each category name using Blockly's messaging format (eg: `<category name="%{BKY_CATLOGIC}">`)
This is why when you use the XML from the code demo with the angular sample, without running through the [step the code demo does](https://github.com/google/blockly/blob/master/demos/code/code.js#L432) to replace these encoded messages with their values from Blockly.Msg you end up with the result above.

To get around this, you can either remove these category names and replace them with English strings, if English is the only language you're hoping to support. 
Alternatively, you can do something like what the code demo does and these messages to the Blockly.Msg Object like so:
 
```ts

import { Component, OnInit } from "@angular/core";

import { BlocklyXmlService } from "src/app/_services/blocklyXml.service";
import * as Blockly from "blockly"; // Includes Blockly core, Blockly built-in blocks, the JavaScript generator and the English lang files.

@Component({
  selector: "app-blockly",
  templateUrl: "./blockly.component.html",
  styleUrls: ["./blockly.component.css"],
  providers: [BlocklyXmlService]
})
export class BlocklyComponent implements OnInit {
  public blocklyXml;

  constructor(private blocklyXmlService: BlocklyXmlService) {}

  createBlocklyDiv(xml: string) {
    const blocklyDiv = document.getElementById("blocklyDiv");

    Blockly.Msg['CATLOGIC'] = 'Logic';
    Blockly.Msg['CATLOOPS'] = 'Loops';
    Blockly.Msg['CATMATH'] = 'Math';
    Blockly.Msg['CATTEXT'] = 'Text';
    Blockly.Msg['CATLISTS'] = 'Lists';
    Blockly.Msg['CATCOLOUR'] = 'Colour';
    Blockly.Msg['CATVARIABLES'] = 'Variables';
    Blockly.Msg['CATFUNCTIONS'] = 'Functions';
    
    if (xml.length > 0) {
      Blockly.inject(blocklyDiv, {
        readOnly: false,
        move: {
          scrollbars: true,
          drag: true,
          wheel: true
        },
        toolbox: xml
      } as Blockly.BlocklyOptions);
    } else {
      console.log("XML data is not here yet.");
    }
  }

  ngOnInit() {
    this.blocklyXmlService
      .getBlocklyXml()
      .subscribe(xmlString => this.createBlocklyDiv(xmlString));
  }
}
```

**Comment by christopherallanperry** [comment]
Sam,

This morning, I've been looking at your suggestions and have had mixed success.

For the messaging format, I've gone with using English strings, that will work fine for our target audience at this point in time.

For the workspace conflict, I've been having less success. Using a component set up as you've shown above, I get the following:

1. -ve: The grid and workspace controls (zoom in/out, zoom reset) are missing
2. -ve: The tabs for the code languages are inaccessible
3. The console shows an `Uncaught TypeError: Cannot read property 'outerHTML' of null`
3. +ve: Blocks placed on the workspace can be receive focus and can be repositioned

![workspace_no_grid](https://user-images.githubusercontent.com/10407622/73171277-4c09d000-40f8-11ea-8685-3ceb16dd69b0.PNG)

![code_js_405_error](https://user-images.githubusercontent.com/10407622/73171896-b2432280-40f9-11ea-9b28-4df3013faa82.PNG)

If I add the line `    blocklyDiv.innerHTML = xml;` back in, I get:

1. +ve: The grid  and controls are visible
2. +ve: The tabs for the code languages  are accessible and showing their respective outputs
3. +ve: No console errors
4. -ve: Once released onto the workspace, blocks cannot receive focus, and cannot be re-positioned. New blocks can be docked with those already on the workspace, but cannot be moved again once the mouse button has been released.

![workspace_no_focus](https://user-images.githubusercontent.com/10407622/73173063-65ad1680-40fc-11ea-9b83-2513d01e737c.PNG)

For context, what I've been asked to look at is whether we could use Blockly as a user friendly way of generating business rules, that we can output into XML to use as a feed for generating [ALFA(XACML)](https://en.wikipedia.org/wiki/ALFA_(XACML)). For that, I need to be able to generate suitable custom blocks, and then output the XML.

I very much appreciate your help on this. Can you offer any guidance on the best way of ending up with the workspace I need to achieve the goal above?

**Comment by christopherallanperry** [comment]
I've been able to resolve the focus/re-positioning issue by setting a negative z-index on `blocklyDiv`. That in combination with putting `blocklyDiv.innerHTML = xml;` back into the `blockly.component.ts` file has got me pretty much where I need to be at this point, though I recognise that there may still be things in the background that need addressing.

I'm happy to close the issue down at this point. Again, many thanks for your help. I've a much better understanding of some of the behaviour of Blockly than I did have.

Thanks
